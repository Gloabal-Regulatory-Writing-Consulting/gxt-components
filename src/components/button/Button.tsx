import React, { ButtonHTMLAttributes, forwardRef, ReactNode } from "react";
import classes from "./Button.module.css";

export interface ButtonProps
  extends Omit<ButtonHTMLAttributes<HTMLButtonElement>, "title"> {
  loading?: boolean;
  disabled?: boolean;
  classNames?: string;
  variant?: "primary" | "secondary" | "negative" | "positive";
  children: ReactNode;
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      loading = false,
      disabled = false,
      classNames = "",
      variant = "primary",
      children,
      ...rest
    }: ButtonProps,
    ref: React.Ref<HTMLButtonElement>,
  ) => {
    const buttonClasses = {
      primary: "primary-button",
      secondary: "secondary-button",
      negative: "negative-button",
      positive: "positive-button",
    };

    const className = `${classes.button} ${classes[buttonClasses[variant]] || classes[buttonClasses.primary]} ${classNames} ${disabled ? "disabled" : ""} `;

    return (
      <button
        ref={ref}
        className={className}
        disabled={loading || disabled}
        {...rest}
      >
        {children}
      </button>
    );
  },
);
Button.displayName = "Button";
export default Button;
