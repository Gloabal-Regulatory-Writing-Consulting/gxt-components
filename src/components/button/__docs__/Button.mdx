import { Canvas, Meta } from "@storybook/blocks";
import * as Button from "./Button.stories.tsx";

<Meta of={Button} title="Button" />

# Button

Button component with different props.

#### Example

<Canvas of={Button.Primary} />

## Usage

```ts
import { Button } from '@gloabal-regulatory-writing-consulting/gxt-components';;

const ExampleButton = () => {
 const handleClick = (event: React.MouseEvent) => {
    // Handle button click
  };

  return (
    <Button
      onClick={handleClick}
      disabled={false}
      variant="primary"
    >
      Click me
    </Button>
  );
};

export default ExampleButton;
```

#### Arguments

- **disabled** -  A boolean indicating whether the button should be disabled.
- **variant** `"primary" | "secondary" | "negative" | "positive"` - The style variant of the button.
- **onClick** - Function called when the button is clicked. It receives a MouseEventHandler for handling the click event on the button element.
