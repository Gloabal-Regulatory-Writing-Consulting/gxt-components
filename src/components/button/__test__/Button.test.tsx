import React from "react";
import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import "jest-styled-components";

import Button from "../Button";

describe("Button component", () => {
  it("Button should not render when disabled and loading", () => {
    render(<Button disabled>Click me</Button>);
    const button = screen.queryByRole("button");
    expect(button).toBeDisabled();
  });
  it("Button should not have type submit if onClick is not defined", () => {
    render(<Button>Click me</Button>);
    const button = screen.getByRole("button");
    expect(button).not.toHaveAttribute("type", "submit");
  });
  it("Button should render title as a child", () => {
    const title = "Click me";
    render(<Button>{title}</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveTextContent(title);
  });
  it("Button should render title as a node", () => {
    const title = <span>Click me</span>;
    render(<Button>{title}</Button>);
    const button = screen.getByRole("button");
    expect(button).toHaveTextContent("Click me");
  });
  it("Button should render when variant is primary", () => {
    const { asFragment } = render(<Button variant="primary">Click Me</Button>);

    expect(asFragment()).toMatchSnapshot("PrimaryButton");
  });
  it("Button should render when variant is secondary", () => {
    const { asFragment } = render(
      <Button variant="secondary">Click Me</Button>,
    );

    expect(asFragment()).toMatchSnapshot("SecondaryButton");
  });
  it("Button should render when variant is negative", () => {
    const { asFragment } = render(<Button variant="negative">Click Me</Button>);

    expect(asFragment()).toMatchSnapshot("NegativeButton");
  });
  it("Button should render when variant is positive", () => {
    const { asFragment } = render(<Button variant="positive">Click Me</Button>);

    expect(asFragment()).toMatchSnapshot("PositiveButton");
  });
});
