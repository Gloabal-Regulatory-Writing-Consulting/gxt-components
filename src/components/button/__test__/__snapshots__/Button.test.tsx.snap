// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > Button should render when variant is negative > NegativeButton 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  gap: 0.75rem;
  border: 1px solid transparent;
  box-shadow: 0px 1px 2px 0px rgba(0, 0, 0, 0.05);
  font-family: Mulish;
  font-size: 1rem;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: 0.00119rem;
  padding: 0.5rem 1rem;
  border-radius: 0.25rem;
}

.c1 {
  background: var(--negative-100, #ef4444);
  border-color: var(--negative-100, #ef4444);
  color: var(--system-50, #fff);
}

.c1:hover {
  background: var(--negative-200, #7f1d1d);
  border-color: var(--negative-200, #7f1d1d);
}

.c1:active {
  border: 2px solid var(--negative-200, #7f1d1d);
  background: var(--negative-50, #f87171);
}

.c1:disabled {
  background: var(--neutral-50, #f9fafb);
  border-color: var(--neutral-50, #f9fafb);
  color: var(--neutral-200, #9ca3af);
}

<button
    class="c0 c1"
  >
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button component > Button should render when variant is positive > PositiveButton 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  gap: 0.75rem;
  border: 1px solid transparent;
  box-shadow: 0px 1px 2px 0px rgba(0, 0, 0, 0.05);
  font-family: Mulish;
  font-size: 1rem;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: 0.00119rem;
  padding: 0.5rem 1rem;
  border-radius: 0.25rem;
}

.c1 {
  background: var(--positive-100, #059669);
  border-color: var(--positive-100, #059669);
  color: var(--system-50, #fff);
}

.c1:hover {
  background: var(--positive-200, #065f46);
  border-color: var(--positive-200, #065f46);
}

.c1:active {
  border: 2px solid var(--positive-200, #065f46);
  background: var(--positive-50, #34d399);
  color: var(--positive-200, #065f46);
}

.c1:disabled {
  background: var(--neutral-50, #f9fafb);
  border-color: var(--neutral-50, #f9fafb);
  color: var(--neutral-200, #9ca3af);
}

<button
    class="c0 c1"
  >
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button component > Button should render when variant is primary > PrimaryButton 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  gap: 0.75rem;
  border: 1px solid transparent;
  box-shadow: 0px 1px 2px 0px rgba(0, 0, 0, 0.05);
  font-family: Mulish;
  font-size: 1rem;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: 0.00119rem;
  padding: 0.5rem 1rem;
  border-radius: 0.25rem;
}

.c1 {
  background: var(--primary-200, #177ba6);
  border-color: var(--primary-200, #177ba6);
  color: var(--system-50, #fff);
}

.c1:hover {
  font-weight: 700;
  background: var(--primary-300, #115873);
  border-color: var(--primary-300, #115873);
  color: var(--system-50, #fff);
}

.c1:active {
  border: 2px solid var(--primary-200, #177ba6);
  background: var(--primary-50, #2aace2);
  color: var(--system-50, #fff);
}

.c1:disabled {
  background: var(--neutral-50, #f9fafb);
  border-color: var(--neutral-50, #f9fafb);
  color: var(--neutral-50, #9ca3af);
}

<button
    class="c0 c1"
  >
    Click Me
  </button>
</DocumentFragment>
`;

exports[`Button component > Button should render when variant is secondary > SecondaryButton 1`] = `
<DocumentFragment>
  .c0 {
  display: inline-flex;
  justify-content: center;
  align-items: center;
  gap: 0.75rem;
  border: 1px solid transparent;
  box-shadow: 0px 1px 2px 0px rgba(0, 0, 0, 0.05);
  font-family: Mulish;
  font-size: 1rem;
  font-style: normal;
  font-weight: 700;
  line-height: 150%;
  letter-spacing: 0.00119rem;
  padding: 0.5rem 1rem;
  border-radius: 0.25rem;
}

.c1 {
  border: 1px solid var(--neutral-50, #9ca3af);
  background: var(--system-50, #fff);
  color: var(--primary-200, #177ba6);
}

.c1:hover {
  border: 1px solid var(--neutral-100, #1c99ce);
  color: var(--primary-200, #177ba6);
}

.c1:active {
  border: 2px solid var(--primary-200, #177ba6);
}

.c1:disabled {
  background: var(--neutral-50, #f9fafb);
  border-color: var(--neutral-50, #f9fafb);
  color: var(--neutral-200, #9ca3af);
}

<button
    class="c0 c1"
  >
    Click Me
  </button>
</DocumentFragment>
`;
